### PR TITLE
[Merged by Bors] - Take DirectionalLight's GlobalTransform into account when calculating shadow map volume (not just direction)

### LINKED ISSUE #5789: Moveable DirectionalLights
## What problem does this solve or what need does it fill?

`DirectionalLight` shadow casting is tied to the world origin.
If you move a camera too far away from the origin, shadows disappear.

## What solution would you like?

Allow the `Transform` component of any entity with a `DirectionalLight` to apply to its shadow casting.
Then, users could update the transform as the camera moves around the world.

## What alternative(s) have you considered?

You might be able to update the frustrum of the `DirectionalLight` manually, but I was unable to get this to work.

### PR DESCRIPTION
**Pull request by coreh**
# Objective

This PR fixes #5789, by enabling movable (and scalable) directional light shadow volumes.

## Solution

This PR changes `ExtractedDirectionalLight` to hold a copy of the `DirectionalLight` entity's `GlobalTransform`, instead of just a `direction` vector. This allows the shadow map volume (as defined by the light's `shadow_projection` field) to be transformed honoring translation _and_ scale transforms, and not just rotation.

It also augments the texel size calculation (used to determine the `shadow_normal_bias`) so that it now takes into account the upper bound of the x/y/z scale of the `GlobalTransform`.

This change makes the directional light extraction code more consistent with point and spot lights (that already use `transform`), and allows easily moving and scaling the shadow volume along with a player entity based on camera distance/angle, immediately enabling more real world use cases until we have a more sophisticated adaptive implementation, such as the one described in #3629.

**Note:** While it was previously possible to update the projection achieving a similar effect, depending on the light direction and distance to the origin, the fact that the shadow map camera was always positioned at the origin with a hardcoded `Vec3::Y` up value meant you would get sub-optimal or inconsistent/incorrect results.

---

### Changed

- `DirectionalLight` shadow volumes now honor translation and scale transforms

## Migration Guide

- If your directional lights were positioned at the origin and not scaled (the default, most common scenario) no changes are needed on your part; it just works as before;
- If you previously had a system for dynamically updating directional light shadow projections, you might now be able to simplify your code by updating the directional light entity's transform instead;
- In the unlikely scenario that a scene with directional lights that previously rendered shadows correctly has missing shadows, make sure your directional lights are positioned at (0, 0, 0) and are not scaled to a size that's too large or too small.

### CODE REVIEW

File: crates/bevy_pbr/src/render/light.rs [hunk 1]
```diff
@@ -560,15 +560,15 @@ pub fn extract_lights(
                 directional_light.shadow_projection.top
                     - directional_light.shadow_projection.bottom,
             );
-        let directional_light_texel_size =
-            largest_dimension / directional_light_shadow_map.size as f32;
+        let directional_light_texel_size = transform.radius_vec3a(Vec3A::ONE) * largest_dimension
```
Comment (HackerFoo) [inline crates/bevy_pbr/src/render/light.rs:563]: Shouldn't the largest dimension be calculated after the transform?
```rust
let directional_light_texel_size = transform.radius_vec3a(Vec3A::new(
  directional_light.shadow_projection.right - directional_light.shadow_projection.left,
  directional_light.shadow_projection.top - directional_light.shadow_projection.bottom,
  0.
)).abs().max_element() / directional_light_shadow_map.size as f32;
```

**Comment by rparrett** [comment]
I can't review the math, but I've been testing with a modified `lighting.rs` that moves the whole scene around and this seems to work.

It would be great if these new behaviors were documented on `DirectionalLight`.

I do agree that even if this isn't the ultimate solution, it would be a huge win over the status quo.

**Comment by rparrett** [comment]
Pinging @ramirezmike

File: crates/bevy_pbr/src/render/light.rs (hunk 1 above)
Comment (coreh) [reply crates/bevy_pbr/src/render/light.rs]: Nice catch. Updated to use this approach, should give more accurate results.

**Comment by coreh** [comment]
> It would be great if these new behaviors were documented on `DirectionalLight`.

Added a section on directional light shadows, documenting how the volume (and shadow maps) work.

**Review by rparrett — APPROVED** [review]
I have been testing in this little "[directional light sandbox](https://gist.github.com/rparrett/fdc16838feba639e7aa1c20e1a334360)" and this seems to be working well.

An interesting note is that prior to this PR, you can modify the translation of a `DirectionalLight`, and even the `lighting` example uses a directional light in a non-zero position. But it's not clear at all what the behavior is supposed to be. When you move the light, the shadow projection is affected, but not in any way that would seemingly make sense to a user. There may be a bug there. If so, this seems to fix it.

I'm really happy about the new docs as well.

**Comment by ramirezmike** [comment]
I checked out this PR locally and updated one of my games to use it and it works like a charm. My game has a system to update the camera's transform based on the player's transform and I just added DirectionalLight to that query filter and now the shadows work regardless of how far I'm from the origin.

I also can't comment too much on the math involved but the parts I grasp make sense and the docs are great 👍

I'll continue testing this out but pretty happy with it so far.

**Review by mockersf — APPROVED** [review]

**Review by alice-i-cecile — APPROVED** [review]
Very nice new docs, fixes an important limitation, reviewers seem to have thoroughly tested this. Added to the milestone.

File: crates/bevy_pbr/src/render/light.rs [hunk 2]
```diff
@@ -550,25 +550,22 @@ pub fn extract_lights(
...
         // https://catlikecoding.com/unity/tutorials/custom-srp/directional-shadows/
-        let largest_dimension = (directional_light.shadow_projection.right
-            - directional_light.shadow_projection.left)
-            .max(
-                directional_light.shadow_projection.top
-                    - directional_light.shadow_projection.bottom,
-            );
-        let directional_light_texel_size =
-            largest_dimension / directional_light_shadow_map.size as f32;
+        let directional_light_texel_size = transform.radius_vec3a(Vec3A::new(
```
Comment (superdump) [inline crates/bevy_pbr/src/render/light.rs:557]: Using the radius (which is calculated as the length of the longest diagonal geometrically speaking) and scaling by sqrt(2) looks incorrect. I think also the previous code using the max dimension and sqrt(2) was incorrect as it assumed square texels which wouldn't have been the case if right - left was not the same as top - bottom.

I think using this radius and removing the factor of sqrt(2) further down would be correct.

**Review by superdump — CHANGES_REQUESTED** [review]

File: crates/bevy_pbr/src/render/light.rs (hunk 2 above)
Comment (coreh) [reply crates/bevy_pbr/src/render/light.rs:557]: Oh, that makes sense. I was originally calculating a scale transform radius and just multiplying the existing texel size value by it. As per @HackerFoo I changed it to do the "proper" calculation in the transformed space, but didn't think of removing the $\sqrt{2}$ for the square's diagonal below.

Fixed.

**Comment by rparrett** [comment]
@superdump Do we need to adjust `DEFAULT_SHADOW_NORMAL_BIAS` now? Playing in the `shadow_biases` example, it seems like an equivalent value to the current default would now be ~`0.9`

**Review by cart — APPROVED** [review]
Code looks good, but I'm conflicted on this one. This is a useful intermediate step and solves a real problem, but it is also training people to think about `DirectionalLights` in a non-standard way (and write code that operates under that assumption). Everywhere else in the ecosystem (ex: Godot, Blender, etc), DirectionalLight position has no bearing on how light / shadows work. I would have preferred a solution that maintains that mental model, but I also understand that the approach in this PR is way more straightforward to implement.

That being said, theres enough positive sentiment here to get me on board. This just increases the pressure to get shadow cascades / view frustum fitting working, which are sorely needed already!

**Comment by superdump** [comment]
@cart - the approach that I implemented a while back was to create an oriented bounding box around the view frustum of the camera and adjust the light's shadow projection based on that. I abandoned it because the shadow map quality quickly degrades with the size of the frustum.

**Comment by coreh** [comment]
Merged and fixed the conflicts. The system now seems to split `ShadowView` and `GpuDirectionalLight` extraction in two separate loops, I believe from #6066. That actually paired nicely with this, the code is overall a bit cleaner.

I tested it and things seem to be working properly with multiple directional lights.

> This is a useful intermediate step and solves a real problem, but it is also training people to think about DirectionalLights in a non-standard way (and write code that operates under that assumption).

😕 Yeah. Please let me know if you'd like me to add a note to the docs explicitly stating that things might change in the future.

> Do we need to adjust DEFAULT_SHADOW_NORMAL_BIAS now? Playing in the shadow_biases example, it seems like an equivalent value to the current default would now be ~0.9

Just tried out the example, is it really supposed to look like this?

image

Tried it out in `main` and it looks the same, so it probably isn't a result of this PR:

image

Or is that what you mean? That we need to tweak the values somehow so that it _doesn't_ look like this?

**Comment by rparrett** [comment]
That example was used to determine the current values for the default shadow bias constants.

I believe the idea is that you use the keys to adjust the biases individually until the shadows don't look bad and then use that value for the default.

But that's what I was seeking clarity from superdump about.

**Comment by superdump** [comment]
> @superdump Do we need to adjust `DEFAULT_SHADOW_NORMAL_BIAS` now? Playing in the `shadow_biases` example, it seems like an equivalent value to the current default would now be ~`0.9`

Why does the normal bias need changing? What changed to need to change it? 🙂 I need to look more closely...

**Comment by rparrett** [comment]
> Why does the normal bias need changing? What changed to need to change it? 🙂 I need to look more closely...

The `sqrt` business, I think.

**Comment by superdump** [comment]
@coreh  yeah the shadow biases example is meant to start off looking bad with lots of shadow artifacts so you know what they look like and can learn how the biases impact them.

**Comment by superdump** [comment]
> > Why does the normal bias need changing? What changed to need to change it? 🙂 I need to look more closely...
> 
> The `sqrt` business, I think.

We changed from sqrt(2) * max(x, y) to sqrt(x^2 + y^2) practically I think. The sqrt(2) came from assuming square fragment aspect ratio so e.g. x == y and then sqrt(x^2 + x^2) == sqrt(2 * x^2) == sqrt(2) * x == sqrt(2) * max(x, x). I feel like that shouldn't have had a significant impact in default cases. Perhaps it or something else I missed does though.

**Comment by rparrett** [comment]
The current `0.6` default still seems like a good value in the `lighting` example. Is it expected for users to need to tune these biases when the projection shape changes?

I just noticed the different behavior in `shadow_biases` where it seemed likely the current defaults were determined (because pre-this-PR the current defaults perfectly eliminate artifacts there) and thought they might need to be updated.

Really don't want to derail this if it's very unlikely to affect normal users though.

**Comment by cart** [comment]
I'm down to merge this. Changing shadow bias behavior (for a given value) should be avoided unless absolutely necessary as artists will carefully tune the value to their specific scenes. It seems like doing it here _was_ necessary though. And changing the _default_ value is less destructive than changing the algorithm, so I'm happy to change that later if we later find a better value.

**Pull request closed without merging**